{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b9aec1a6",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting optuna\n",
      "  Using cached optuna-3.4.0-py3-none-any.whl (409 kB)\n",
      "Requirement already satisfied: transformers in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (4.35.0)\n",
      "Collecting datasets\n",
      "  Using cached datasets-2.15.0-py3-none-any.whl (521 kB)\n",
      "Collecting accelerate\n",
      "  Using cached accelerate-0.24.1-py3-none-any.whl (261 kB)\n",
      "Collecting evaluate\n",
      "  Using cached evaluate-0.4.1-py3-none-any.whl (84 kB)\n",
      "Collecting alembic>=1.5.0\n",
      "  Using cached alembic-1.12.1-py3-none-any.whl (226 kB)\n",
      "Collecting colorlog\n",
      "  Using cached colorlog-6.7.0-py2.py3-none-any.whl (11 kB)\n",
      "Requirement already satisfied: tqdm in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (4.66.1)\n",
      "Requirement already satisfied: PyYAML in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (6.0)\n",
      "Requirement already satisfied: sqlalchemy>=1.3.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (1.4.32)\n",
      "Requirement already satisfied: numpy in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (1.26.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (21.3)\n",
      "Requirement already satisfied: tokenizers<0.15,>=0.14 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from transformers) (0.14.1)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from transformers) (2023.8.8)\n",
      "Requirement already satisfied: filelock in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from transformers) (3.13.1)\n",
      "Requirement already satisfied: requests in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from transformers) (2.27.1)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.16.4 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from transformers) (0.17.3)\n",
      "Requirement already satisfied: safetensors>=0.3.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from transformers) (0.4.0)\n",
      "Requirement already satisfied: aiohttp in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from datasets) (3.8.1)\n",
      "Requirement already satisfied: fsspec[http]<=2023.10.0,>=2023.1.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from datasets) (2023.10.0)\n",
      "Collecting huggingface-hub<1.0,>=0.16.4\n",
      "  Using cached huggingface_hub-0.19.4-py3-none-any.whl (311 kB)\n",
      "Collecting pyarrow-hotfix\n",
      "  Using cached pyarrow_hotfix-0.5-py3-none-any.whl (7.8 kB)\n",
      "Requirement already satisfied: pandas in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from datasets) (1.4.2)\n",
      "Collecting xxhash\n",
      "  Using cached xxhash-3.4.1-cp39-cp39-macosx_10_9_x86_64.whl (31 kB)\n",
      "Collecting multiprocess\n",
      "  Using cached multiprocess-0.70.15-py39-none-any.whl (133 kB)\n",
      "Collecting dill<0.3.8,>=0.3.0\n",
      "  Using cached dill-0.3.7-py3-none-any.whl (115 kB)\n",
      "Collecting pyarrow>=8.0.0\n",
      "  Using cached pyarrow-14.0.1-cp39-cp39-macosx_10_14_x86_64.whl (26.9 MB)\n",
      "Requirement already satisfied: torch>=1.10.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from accelerate) (2.1.0)\n",
      "Requirement already satisfied: psutil in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from accelerate) (5.8.0)\n",
      "Collecting responses<0.19\n",
      "  Using cached responses-0.18.0-py3-none-any.whl (38 kB)\n",
      "Collecting Mako\n",
      "  Using cached Mako-1.3.0-py3-none-any.whl (78 kB)\n",
      "Requirement already satisfied: typing-extensions>=4 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from alembic>=1.5.0->optuna) (4.1.1)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets) (5.2.0)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets) (4.0.1)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets) (1.6.3)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets) (21.4.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets) (1.2.0)\n",
      "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets) (2.0.4)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets) (1.2.0)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from packaging>=20.0->optuna) (3.0.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests->transformers) (1.26.9)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests->transformers) (3.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests->transformers) (2021.10.8)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from sqlalchemy>=1.3.0->optuna) (1.1.1)\n",
      "Collecting tokenizers<0.15,>=0.14\n",
      "  Using cached tokenizers-0.14.0-cp39-cp39-macosx_10_7_x86_64.whl (2.5 MB)\n",
      "INFO: pip is looking at multiple versions of greenlet to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting greenlet!=0.4.17\n",
      "  Using cached greenlet-3.0.1-cp39-cp39-macosx_10_9_universal2.whl (260 kB)\n",
      "  Using cached greenlet-3.0.0-cp39-cp39-macosx_10_9_universal2.whl (260 kB)\n",
      "  Using cached greenlet-2.0.2.tar.gz (164 kB)\n",
      "  Using cached greenlet-2.0.1-cp39-cp39-macosx_10_15_x86_64.whl (203 kB)\n",
      "  Using cached greenlet-2.0.0.post0-cp39-cp39-macosx_10_15_x86_64.whl (200 kB)\n",
      "  Using cached greenlet-2.0.0-cp39-cp39-macosx_10_15_x86_64.whl (200 kB)\n",
      "  Using cached greenlet-1.1.3.post0-cp39-cp39-macosx_10_15_x86_64.whl (93 kB)\n",
      "INFO: pip is looking at multiple versions of greenlet to determine which version is compatible with other requirements. This could take a while.\n",
      "  Using cached greenlet-1.1.3-cp39-cp39-macosx_10_15_x86_64.whl (93 kB)\n",
      "  Using cached greenlet-1.1.2-cp39-cp39-macosx_10_14_x86_64.whl (92 kB)\n",
      "  Using cached greenlet-1.1.1-cp39-cp39-macosx_10_14_x86_64.whl (87 kB)\n",
      "  Using cached greenlet-1.1.0-cp39-cp39-macosx_10_14_x86_64.whl (87 kB)\n",
      "  Using cached greenlet-1.0.0-cp39-cp39-macosx_10_14_x86_64.whl (86 kB)\n",
      "INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking\n",
      "  Using cached greenlet-0.4.16.tar.gz (60 kB)\n",
      "  Using cached greenlet-0.4.15.tar.gz (59 kB)\n",
      "  Using cached greenlet-0.4.14.tar.gz (59 kB)\n",
      "  Using cached greenlet-0.4.13.tar.gz (58 kB)\n",
      "  Using cached greenlet-0.4.12.tar.gz (57 kB)\n",
      "  Using cached greenlet-0.4.11.zip (82 kB)\n",
      "  Using cached greenlet-0.4.10.zip (82 kB)\n",
      "  Using cached greenlet-0.4.9.zip (78 kB)\n",
      "  Using cached greenlet-0.4.7.zip (77 kB)\n",
      "  Using cached greenlet-0.4.6.zip (77 kB)\n",
      "  Using cached greenlet-0.4.5.zip (77 kB)\n",
      "  Using cached greenlet-0.4.4.zip (76 kB)\n",
      "  Using cached greenlet-0.4.3.zip (76 kB)\n",
      "  Using cached greenlet-0.4.2.zip (74 kB)\n",
      "  Using cached greenlet-0.4.1.zip (75 kB)\n",
      "  Using cached greenlet-0.4.0.zip (72 kB)\n",
      "  Using cached greenlet-0.3.4.zip (63 kB)\n",
      "  Using cached greenlet-0.3.3.zip (65 kB)\n",
      "  Using cached greenlet-0.3.2.zip (50 kB)\n",
      "  Using cached greenlet-0.3.1.tar.gz (39 kB)\n",
      "  Using cached greenlet-0.3.tar.gz (37 kB)\n",
      "  Using cached greenlet-0.2.tar.gz (22 kB)\n",
      "\u001b[31m    ERROR: Command errored out with exit status 1:\n",
      "     command: /Users/dannyliu/opt/anaconda3/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '\"'\"'/private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_2b0e28c5e21748c79b5f12326a053bbc/setup.py'\"'\"'; __file__='\"'\"'/private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_2b0e28c5e21748c79b5f12326a053bbc/setup.py'\"'\"';f = getattr(tokenize, '\"'\"'open'\"'\"', open)(__file__) if os.path.exists(__file__) else io.StringIO('\"'\"'from setuptools import setup; setup()'\"'\"');code = f.read().replace('\"'\"'\\r\\n'\"'\"', '\"'\"'\\n'\"'\"');f.close();exec(compile(code, __file__, '\"'\"'exec'\"'\"'))' egg_info --egg-base /private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-pip-egg-info-2badp0_s\n",
      "         cwd: /private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_2b0e28c5e21748c79b5f12326a053bbc/\n",
      "    Complete output (8 lines):\n",
      "    Traceback (most recent call last):\n",
      "      File \"<string>\", line 1, in <module>\n",
      "      File \"/private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_2b0e28c5e21748c79b5f12326a053bbc/setup.py\", line 3, in <module>\n",
      "        import ez_setup\n",
      "      File \"/private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_2b0e28c5e21748c79b5f12326a053bbc/ez_setup/__init__.py\", line 98\n",
      "        except pkg_resources.VersionConflict, e:\n",
      "                                            ^\n",
      "    SyntaxError: invalid syntax\n",
      "    ----------------------------------------\u001b[0m\n",
      "\u001b[33mWARNING: Discarding https://files.pythonhosted.org/packages/72/db/0d1a944ba1517256e8b1a080a90297569d8662442631a6b05f8a1c584556/greenlet-0.2.tar.gz#sha256=d1919d283f7e7a2cbb474c620efb703811408026f8c6b25f923acd8b49deebaa (from https://pypi.org/simple/greenlet/). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.\u001b[0m\n",
      "  Using cached greenlet-0.1.tar.gz (18 kB)\n",
      "\u001b[31m    ERROR: Command errored out with exit status 1:\n",
      "     command: /Users/dannyliu/opt/anaconda3/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '\"'\"'/private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_3eb6b525afb74e1085aaabb26a376953/setup.py'\"'\"'; __file__='\"'\"'/private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_3eb6b525afb74e1085aaabb26a376953/setup.py'\"'\"';f = getattr(tokenize, '\"'\"'open'\"'\"', open)(__file__) if os.path.exists(__file__) else io.StringIO('\"'\"'from setuptools import setup; setup()'\"'\"');code = f.read().replace('\"'\"'\\r\\n'\"'\"', '\"'\"'\\n'\"'\"');f.close();exec(compile(code, __file__, '\"'\"'exec'\"'\"'))' egg_info --egg-base /private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-pip-egg-info-an0mhrul\n",
      "         cwd: /private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_3eb6b525afb74e1085aaabb26a376953/\n",
      "    Complete output (8 lines):\n",
      "    Traceback (most recent call last):\n",
      "      File \"<string>\", line 1, in <module>\n",
      "      File \"/private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_3eb6b525afb74e1085aaabb26a376953/setup.py\", line 3, in <module>\n",
      "        import ez_setup\n",
      "      File \"/private/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/pip-install-0b9hakt0/greenlet_3eb6b525afb74e1085aaabb26a376953/ez_setup/__init__.py\", line 176\n",
      "        print \"Setuptools version\",version,\"or greater has been installed.\"\n",
      "              ^\n",
      "    SyntaxError: Missing parentheses in call to 'print'. Did you mean print(\"Setuptools version\",version,\"or greater has been installed.\")?\n",
      "    ----------------------------------------\u001b[0m\n",
      "\u001b[33mWARNING: Discarding https://files.pythonhosted.org/packages/8e/6d/59d9a80285c4fef1254d4f589dda87d593e7316082ff94ad2f277eafca22/greenlet-0.1.tar.gz#sha256=5d71dd7e5acb64333c387fbf215f0019ffcdaeab46bb4d30f9ff6ccc94e41710 (from https://pypi.org/simple/greenlet/). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.\u001b[0m\n",
      "INFO: pip is looking at multiple versions of sqlalchemy to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting sqlalchemy>=1.3.0\n",
      "  Using cached SQLAlchemy-2.0.23-cp39-cp39-macosx_10_9_x86_64.whl (2.1 MB)\n",
      "Collecting typing-extensions>=4\n",
      "  Using cached typing_extensions-4.8.0-py3-none-any.whl (31 kB)\n",
      "Collecting sqlalchemy>=1.3.0\n",
      "  Using cached SQLAlchemy-2.0.22-cp39-cp39-macosx_10_9_x86_64.whl (2.1 MB)\n",
      "  Using cached SQLAlchemy-2.0.21-cp39-cp39-macosx_10_9_x86_64.whl (2.1 MB)\n",
      "  Using cached SQLAlchemy-2.0.20-cp39-cp39-macosx_10_9_x86_64.whl (2.1 MB)\n",
      "  Using cached SQLAlchemy-2.0.19-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.18-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.17-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "INFO: pip is looking at multiple versions of sqlalchemy to determine which version is compatible with other requirements. This could take a while.\n",
      "  Using cached SQLAlchemy-2.0.16-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.15-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.14-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.13-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.12-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking\n",
      "  Using cached SQLAlchemy-2.0.11-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.10-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.9-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.8-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.7-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.6-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.5.post1-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.4-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.3-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.2-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.1-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-2.0.0-cp39-cp39-macosx_10_9_x86_64.whl (2.0 MB)\n",
      "  Using cached SQLAlchemy-1.4.50.tar.gz (8.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.49.tar.gz (8.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.48.tar.gz (8.4 MB)\n",
      "  Using cached SQLAlchemy-1.4.47.tar.gz (8.6 MB)\n",
      "  Using cached SQLAlchemy-1.4.46.tar.gz (8.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.45-cp39-cp39-macosx_10_15_x86_64.whl (1.6 MB)\n",
      "  Using cached SQLAlchemy-1.4.44-cp39-cp39-macosx_10_15_x86_64.whl (1.6 MB)\n",
      "  Using cached SQLAlchemy-1.4.43-cp39-cp39-macosx_10_15_x86_64.whl (1.6 MB)\n",
      "  Using cached SQLAlchemy-1.4.42-cp39-cp39-macosx_10_15_x86_64.whl (1.6 MB)\n",
      "  Using cached SQLAlchemy-1.4.41-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.40-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.39-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.38-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.37-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.36-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.35-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.32-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.31-cp39-cp39-macosx_10_15_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.30-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.29-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.28-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.27-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.26-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.25-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.24-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.23-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.22-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.21-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.20-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.19-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.18-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.17-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.16-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.15-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.14-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.13-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.12-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.11-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.10-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.9-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.8-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.7-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.6-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.5-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.4-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.3-cp39-cp39-macosx_10_14_x86_64.whl (1.5 MB)\n",
      "  Using cached SQLAlchemy-1.4.2-cp39-cp39-macosx_10_14_x86_64.whl (1.4 MB)\n",
      "  Using cached SQLAlchemy-1.4.1-cp39-cp39-macosx_10_14_x86_64.whl (1.4 MB)\n",
      "  Using cached SQLAlchemy-1.4.0-cp39-cp39-macosx_10_14_x86_64.whl (1.4 MB)\n",
      "  Using cached SQLAlchemy-1.3.24-cp39-cp39-macosx_10_14_x86_64.whl (1.2 MB)\n",
      "  Using cached SQLAlchemy-1.3.23-cp39-cp39-macosx_10_14_x86_64.whl (1.2 MB)\n",
      "  Using cached SQLAlchemy-1.3.22-cp39-cp39-macosx_10_14_x86_64.whl (1.2 MB)\n",
      "  Using cached SQLAlchemy-1.3.21-cp39-cp39-macosx_10_14_x86_64.whl (1.2 MB)\n",
      "  Using cached SQLAlchemy-1.3.20-cp39-cp39-macosx_10_14_x86_64.whl (1.2 MB)\n",
      "  Using cached SQLAlchemy-1.3.19.tar.gz (6.2 MB)\n",
      "  Using cached SQLAlchemy-1.3.18.tar.gz (6.0 MB)\n",
      "  Using cached SQLAlchemy-1.3.17.tar.gz (6.0 MB)\n",
      "  Using cached SQLAlchemy-1.3.16.tar.gz (6.1 MB)\n",
      "  Using cached SQLAlchemy-1.3.15.tar.gz (6.1 MB)\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h    Preparing wheel metadata ... \u001b[?25ldone\n",
      "\u001b[?25h  Using cached SQLAlchemy-1.3.14.tar.gz (6.1 MB)\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h    Preparing wheel metadata ... \u001b[?25ldone\n",
      "\u001b[?25h  Using cached SQLAlchemy-1.3.13.tar.gz (6.0 MB)\n",
      "  Using cached SQLAlchemy-1.3.12.tar.gz (6.0 MB)\n",
      "  Using cached SQLAlchemy-1.3.11.tar.gz (6.0 MB)\n",
      "  Using cached SQLAlchemy-1.3.10.tar.gz (6.0 MB)\n",
      "  Using cached SQLAlchemy-1.3.9.tar.gz (6.0 MB)\n",
      "  Using cached SQLAlchemy-1.3.8.tar.gz (5.9 MB)\n",
      "  Using cached SQLAlchemy-1.3.7.tar.gz (5.9 MB)\n",
      "  Using cached SQLAlchemy-1.3.6.tar.gz (5.9 MB)\n",
      "  Using cached SQLAlchemy-1.3.5.tar.gz (5.9 MB)\n",
      "  Using cached SQLAlchemy-1.3.4.tar.gz (5.9 MB)\n",
      "  Using cached SQLAlchemy-1.3.3.tar.gz (5.9 MB)\n",
      "  Using cached SQLAlchemy-1.3.2.tar.gz (5.9 MB)\n",
      "  Using cached SQLAlchemy-1.3.1.tar.gz (5.9 MB)\n",
      "  Using cached SQLAlchemy-1.3.0.tar.gz (5.9 MB)\n",
      "INFO: pip is looking at multiple versions of safetensors to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting safetensors>=0.3.1\n",
      "  Using cached safetensors-0.4.0-cp39-cp39-macosx_10_7_x86_64.whl (439 kB)\n",
      "  Using cached safetensors-0.3.3-cp39-cp39-macosx_10_11_x86_64.whl (404 kB)\n",
      "  Using cached safetensors-0.3.2-cp39-cp39-macosx_10_11_x86_64.whl (403 kB)\n",
      "  Using cached safetensors-0.3.1-cp39-cp39-macosx_10_11_x86_64.whl (400 kB)\n",
      "INFO: pip is looking at multiple versions of responses to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting responses<0.19\n",
      "  Using cached responses-0.17.0-py2.py3-none-any.whl (38 kB)\n",
      "Requirement already satisfied: six in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from responses<0.19->evaluate) (1.16.0)\n",
      "INFO: pip is looking at multiple versions of safetensors to determine which version is compatible with other requirements. This could take a while.\n",
      "  Using cached responses-0.16.0-py2.py3-none-any.whl (35 kB)\n",
      "INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking\n",
      "  Using cached responses-0.15.0-py2.py3-none-any.whl (32 kB)\n",
      "  Using cached responses-0.14.0-py2.py3-none-any.whl (27 kB)\n",
      "  Using cached responses-0.13.4-py2.py3-none-any.whl (25 kB)\n",
      "  Using cached responses-0.13.3-py2.py3-none-any.whl (25 kB)\n",
      "  Using cached responses-0.13.2-py2.py3-none-any.whl (25 kB)\n",
      "INFO: pip is looking at multiple versions of responses to determine which version is compatible with other requirements. This could take a while.\n",
      "  Using cached responses-0.13.1-py2.py3-none-any.whl (25 kB)\n",
      "  Using cached responses-0.13.0-py2.py3-none-any.whl (23 kB)\n",
      "  Using cached responses-0.12.1-py2.py3-none-any.whl (16 kB)\n",
      "  Using cached responses-0.12.0-py2.py3-none-any.whl (15 kB)\n",
      "  Using cached responses-0.11.0-py2.py3-none-any.whl (15 kB)\n",
      "INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking\n",
      "  Using cached responses-0.10.16-py2.py3-none-any.whl (15 kB)\n",
      "  Using cached responses-0.10.15-py2.py3-none-any.whl (15 kB)\n",
      "  Using cached responses-0.10.14-py2.py3-none-any.whl (15 kB)\n",
      "  Using cached responses-0.10.12-py2.py3-none-any.whl (15 kB)\n",
      "  Using cached responses-0.10.11-py2.py3-none-any.whl (15 kB)\n",
      "  Using cached responses-0.10.9-py2.py3-none-any.whl (14 kB)\n",
      "  Using cached responses-0.10.8-py2.py3-none-any.whl (14 kB)\n",
      "  Using cached responses-0.10.7-py2.py3-none-any.whl (14 kB)\n",
      "  Using cached responses-0.10.6-py2.py3-none-any.whl (13 kB)\n",
      "  Using cached responses-0.10.5-py2.py3-none-any.whl (17 kB)\n",
      "  Using cached responses-0.10.4-py2.py3-none-any.whl (17 kB)\n",
      "  Using cached responses-0.10.3-py2.py3-none-any.whl (17 kB)\n",
      "  Using cached responses-0.10.2-py2.py3-none-any.whl (17 kB)\n",
      "  Using cached responses-0.10.1-py2.py3-none-any.whl (17 kB)\n",
      "Collecting biscuits\n",
      "  Using cached biscuits-0.3.1.tar.gz (93 kB)\n",
      "Collecting responses<0.19\n",
      "  Using cached responses-0.10.0.tar.gz (20 kB)\n",
      "  Using cached responses-0.9.0-py2.py3-none-any.whl (12 kB)\n",
      "Collecting cookies\n",
      "  Using cached cookies-2.2.1-py2.py3-none-any.whl (44 kB)\n",
      "Collecting responses<0.19\n",
      "  Using cached responses-0.8.1-py2.py3-none-any.whl (15 kB)\n",
      "  Using cached responses-0.8.0-py2.py3-none-any.whl (14 kB)\n",
      "  Using cached responses-0.7.0-py2.py3-none-any.whl (13 kB)\n",
      "  Using cached responses-0.6.2-py2.py3-none-any.whl (13 kB)\n",
      "  Using cached responses-0.6.1-py2.py3-none-any.whl (12 kB)\n",
      "  Using cached responses-0.6.0-py2.py3-none-any.whl (12 kB)\n",
      "  Using cached responses-0.5.1-py2.py3-none-any.whl (7.5 kB)\n",
      "  Using cached responses-0.5.0-py2.py3-none-any.whl (7.7 kB)\n",
      "  Using cached responses-0.4.0.tar.gz (5.5 kB)\n",
      "Requirement already satisfied: mock in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from responses<0.19->evaluate) (4.0.3)\n",
      "  Using cached responses-0.3.0.tar.gz (5.1 kB)\n",
      "  Using cached responses-0.2.2.tar.gz (3.7 kB)\n",
      "  Using cached responses-0.2.1.tar.gz (3.7 kB)\n",
      "  Using cached responses-0.2.0.tar.gz (3.7 kB)\n",
      "  Using cached responses-0.1.1.tar.gz (3.4 kB)\n",
      "  Using cached responses-0.1.0.tar.gz (3.2 kB)\n",
      "INFO: pip is looking at multiple versions of idna to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting idna<4,>=2.5\n",
      "  Using cached idna-3.4-py3-none-any.whl (61 kB)\n",
      "\u001b[31mERROR: Exception:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/pip/_internal/cli/base_command.py\", line 173, in _main\n",
      "    status = self.run(options, args)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/pip/_internal/cli/req_command.py\", line 203, in wrapper\n",
      "    return func(self, options, args)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/pip/_internal/commands/install.py\", line 315, in run\n",
      "    requirement_set = resolver.resolve(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py\", line 94, in resolve\n",
      "    result = self._result = resolver.resolve(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py\", line 472, in resolve\n",
      "    state = resolution.resolve(requirements, max_rounds=max_rounds)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py\", line 383, in resolve\n",
      "    raise ResolutionTooDeep(max_rounds)\n",
      "pip._vendor.resolvelib.resolvers.ResolutionTooDeep: 2000000\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install optuna transformers accelerate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "91d47da0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting evaluate\n",
      "  Using cached evaluate-0.4.1-py3-none-any.whl (84 kB)\n",
      "Collecting responses<0.19\n",
      "  Using cached responses-0.18.0-py3-none-any.whl (38 kB)\n",
      "Requirement already satisfied: tqdm>=4.62.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from evaluate) (4.66.1)\n",
      "Requirement already satisfied: fsspec[http]>=2021.05.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from evaluate) (2023.10.0)\n",
      "Requirement already satisfied: numpy>=1.17 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from evaluate) (1.26.0)\n",
      "Collecting xxhash\n",
      "  Using cached xxhash-3.4.1-cp39-cp39-macosx_10_9_x86_64.whl (31 kB)\n",
      "Requirement already satisfied: pandas in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from evaluate) (1.4.2)\n",
      "Requirement already satisfied: huggingface-hub>=0.7.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from evaluate) (0.17.3)\n",
      "Requirement already satisfied: packaging in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from evaluate) (21.3)\n",
      "Requirement already satisfied: requests>=2.19.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from evaluate) (2.27.1)\n",
      "Collecting datasets>=2.0.0\n",
      "  Using cached datasets-2.15.0-py3-none-any.whl (521 kB)\n",
      "Collecting multiprocess\n",
      "  Using cached multiprocess-0.70.15-py39-none-any.whl (133 kB)\n",
      "Collecting dill\n",
      "  Using cached dill-0.3.7-py3-none-any.whl (115 kB)\n",
      "Collecting pyarrow-hotfix\n",
      "  Using cached pyarrow_hotfix-0.5-py3-none-any.whl (7.8 kB)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from datasets>=2.0.0->evaluate) (6.0)\n",
      "Collecting pyarrow>=8.0.0\n",
      "  Using cached pyarrow-14.0.1-cp39-cp39-macosx_10_14_x86_64.whl (26.9 MB)\n",
      "Requirement already satisfied: aiohttp in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from datasets>=2.0.0->evaluate) (3.8.1)\n",
      "Collecting huggingface-hub>=0.7.0\n",
      "  Using cached huggingface_hub-0.19.4-py3-none-any.whl (311 kB)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets>=2.0.0->evaluate) (1.6.3)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets>=2.0.0->evaluate) (1.2.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets>=2.0.0->evaluate) (5.2.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets>=2.0.0->evaluate) (1.2.0)\n",
      "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets>=2.0.0->evaluate) (2.0.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets>=2.0.0->evaluate) (4.0.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from aiohttp->datasets>=2.0.0->evaluate) (21.4.0)\n",
      "Requirement already satisfied: typing-extensions>=3.6.5 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from async-timeout<5.0,>=4.0.0a3->aiohttp->datasets>=2.0.0->evaluate) (4.1.1)\n",
      "Requirement already satisfied: filelock in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from huggingface-hub>=0.7.0->evaluate) (3.13.1)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from packaging->evaluate) (3.0.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.19.0->evaluate) (1.26.9)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.19.0->evaluate) (3.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.19.0->evaluate) (2021.10.8)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from pandas->evaluate) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from pandas->evaluate) (2021.3)\n",
      "Requirement already satisfied: six>=1.5 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.8.1->pandas->evaluate) (1.16.0)\n",
      "Installing collected packages: dill, xxhash, pyarrow-hotfix, pyarrow, multiprocess, huggingface-hub, responses, datasets, evaluate\n",
      "  Attempting uninstall: huggingface-hub\n",
      "    Found existing installation: huggingface-hub 0.17.3\n",
      "    Uninstalling huggingface-hub-0.17.3:\n",
      "      Successfully uninstalled huggingface-hub-0.17.3\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "tokenizers 0.14.1 requires huggingface_hub<0.18,>=0.16.4, but you have huggingface-hub 0.19.4 which is incompatible.\u001b[0m\n",
      "Successfully installed datasets-2.15.0 dill-0.3.7 evaluate-0.4.1 huggingface-hub-0.19.4 multiprocess-0.70.15 pyarrow-14.0.1 pyarrow-hotfix-0.5 responses-0.18.0 xxhash-3.4.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install evaluate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4dfbe1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting accelerate\n",
      "  Using cached accelerate-0.24.1-py3-none-any.whl (261 kB)\n",
      "Requirement already satisfied: torch>=1.10.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from accelerate) (2.1.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from accelerate) (21.3)\n",
      "Requirement already satisfied: numpy>=1.17 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from accelerate) (1.26.0)\n",
      "Requirement already satisfied: psutil in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from accelerate) (5.8.0)\n",
      "Requirement already satisfied: pyyaml in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from accelerate) (6.0)\n",
      "Requirement already satisfied: huggingface-hub in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from accelerate) (0.19.4)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from packaging>=20.0->accelerate) (3.0.4)\n",
      "Requirement already satisfied: filelock in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from torch>=1.10.0->accelerate) (3.13.1)\n",
      "Requirement already satisfied: typing-extensions in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from torch>=1.10.0->accelerate) (4.1.1)\n",
      "Requirement already satisfied: sympy in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from torch>=1.10.0->accelerate) (1.10.1)\n",
      "Requirement already satisfied: networkx in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from torch>=1.10.0->accelerate) (2.7.1)\n",
      "Requirement already satisfied: jinja2 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from torch>=1.10.0->accelerate) (2.11.3)\n",
      "Requirement already satisfied: fsspec in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from torch>=1.10.0->accelerate) (2023.10.0)\n",
      "Requirement already satisfied: tqdm>=4.42.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from huggingface-hub->accelerate) (4.66.1)\n",
      "Requirement already satisfied: requests in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from huggingface-hub->accelerate) (2.27.1)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from jinja2->torch>=1.10.0->accelerate) (2.0.1)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests->huggingface-hub->accelerate) (1.26.9)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests->huggingface-hub->accelerate) (3.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests->huggingface-hub->accelerate) (2021.10.8)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from requests->huggingface-hub->accelerate) (2.0.4)\n",
      "Requirement already satisfied: mpmath>=0.19 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from sympy->torch>=1.10.0->accelerate) (1.2.1)\n",
      "Installing collected packages: accelerate\n",
      "Successfully installed accelerate-0.24.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install accelerate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0e1d6e27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting optuna\n",
      "  Using cached optuna-3.4.0-py3-none-any.whl (409 kB)\n",
      "Collecting alembic>=1.5.0\n",
      "  Using cached alembic-1.12.1-py3-none-any.whl (226 kB)\n",
      "Requirement already satisfied: tqdm in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (4.66.1)\n",
      "Requirement already satisfied: numpy in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (1.26.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (21.3)\n",
      "Requirement already satisfied: PyYAML in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (6.0)\n",
      "Requirement already satisfied: sqlalchemy>=1.3.0 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from optuna) (1.4.32)\n",
      "Collecting colorlog\n",
      "  Using cached colorlog-6.7.0-py2.py3-none-any.whl (11 kB)\n",
      "Requirement already satisfied: typing-extensions>=4 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from alembic>=1.5.0->optuna) (4.1.1)\n",
      "Collecting Mako\n",
      "  Using cached Mako-1.3.0-py3-none-any.whl (78 kB)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from packaging>=20.0->optuna) (3.0.4)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from sqlalchemy>=1.3.0->optuna) (1.1.1)\n",
      "Requirement already satisfied: MarkupSafe>=0.9.2 in /Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages (from Mako->alembic>=1.5.0->optuna) (2.0.1)\n",
      "Installing collected packages: Mako, colorlog, alembic, optuna\n",
      "Successfully installed Mako-1.3.0 alembic-1.12.1 colorlog-6.7.0 optuna-3.4.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install optuna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1353f197",
   "metadata": {},
   "outputs": [],
   "source": [
    "from transformers import AutoModelForSequenceClassification, AutoTokenizer\n",
    "import torch, evaluate, accelerate\n",
    "from transformers import TrainingArguments, Trainer\n",
    "import numpy as np\n",
    "import optuna\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7db1034b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mps'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set device to cuda/mps/cpu\n",
    "if torch.cuda.is_available():\n",
    "    device = \"cuda\"\n",
    "elif torch.backends.mps.is_available():\n",
    "    device = \"mps\"\n",
    "else:\n",
    "    device = \"cpu\"\n",
    "device"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a2f1e701",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unexpected exception formatting exception. Falling back to standard exception\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/huggingface_hub/utils/_errors.py\", line 261, in hf_raise_for_status\n",
      "    find.\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/requests/models.py\", line 960, in raise_for_status\n",
      "    raise HTTPError(http_error_msg, response=self)\n",
      "requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://huggingface.co/meta-llama/Llama-2-7b-hf/resolve/main/tokenizer_config.json\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/transformers/utils/hub.py\", line 430, in cached_file\n",
      "    resolved_file = hf_hub_download(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/huggingface_hub/utils/_validators.py\", line 118, in _inner_fn\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/huggingface_hub/file_download.py\", line 1346, in hf_hub_download\n",
      "    commit_hash = revision\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/huggingface_hub/file_download.py\", line 1232, in hf_hub_download\n",
      "    headers = build_hf_headers(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/huggingface_hub/utils/_validators.py\", line 118, in _inner_fn\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/huggingface_hub/file_download.py\", line 1608, in get_hf_file_metadata\n",
      "    - If `False` or `None`, no token is provided.\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/huggingface_hub/utils/_errors.py\", line 277, in hf_raise_for_status\n",
      "huggingface_hub.utils._errors.GatedRepoError: 401 Client Error. (Request ID: Root=1-6556c3cd-4c72d8962244c03f0236a41f;2cec9e4f-0f78-4a4e-b974-65b245bd28bf)\n",
      "\n",
      "Cannot access gated repo for url https://huggingface.co/meta-llama/Llama-2-7b-hf/resolve/main/tokenizer_config.json.\n",
      "Repo model meta-llama/Llama-2-7b-hf is gated. You must be authenticated to access it.\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3369, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/m5/cx4hwql56m5_7gzqnfv0v5q80000gn/T/ipykernel_17700/3198344656.py\", line 2, in <cell line: 2>\n",
      "    tokenizer = AutoTokenizer.from_pretrained(modelname, use_fast=True)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/transformers/models/auto/tokenization_auto.py\", line 718, in from_pretrained\n",
      "    tokenizer_config = get_tokenizer_config(pretrained_model_name_or_path, **kwargs)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/transformers/models/auto/tokenization_auto.py\", line 550, in get_tokenizer_config\n",
      "    resolved_config_file = cached_file(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/transformers/utils/hub.py\", line 445, in cached_file\n",
      "    raise EnvironmentError(\n",
      "OSError: You are trying to access a gated repo.\n",
      "Make sure to request access at https://huggingface.co/meta-llama/Llama-2-7b-hf and pass a token having permission to this repo either by logging in with `huggingface-cli login` or by passing `token=<your_token>`.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 1982, in showtraceback\n",
      "    stb = self.InteractiveTB.structured_traceback(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 1118, in structured_traceback\n",
      "    return FormattedTB.structured_traceback(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 1012, in structured_traceback\n",
      "    return VerboseTB.structured_traceback(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 884, in structured_traceback\n",
      "    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 818, in format_exception_as_a_whole\n",
      "    frames.append(self.format_record(r))\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 736, in format_record\n",
      "    result += ''.join(_format_traceback_lines(frame_info.lines, Colors, self.has_colors, lvals))\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 698, in lines\n",
      "    pieces = self.included_pieces\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 649, in included_pieces\n",
      "    pos = scope_pieces.index(self.executing_piece)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/stack_data/core.py\", line 628, in executing_piece\n",
      "    return only(\n",
      "  File \"/Users/dannyliu/opt/anaconda3/lib/python3.9/site-packages/executing/executing.py\", line 164, in only\n",
      "    raise NotOneValueFound('Expected one value, found 0')\n",
      "executing.executing.NotOneValueFound: Expected one value, found 0\n"
     ]
    }
   ],
   "source": [
    "modelname = \"meta-llama/Llama-2-7b-hf\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(modelname, use_fast=True)\n",
    "model = AutoModelForSequenceClassification.from_pretrained(modelname,\n",
    "                                                            num_labels=2).to(device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cba52cf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
